src/Symfony/Component/Debug/DebugClassLoader.php

```diff
@@ -163,11 +164,21 @@ public function loadClass($class)
                 throw new \RuntimeException(sprintf('Case mismatch between loaded and declared class names: %s vs %s', $class, $name));
             }
 
+            if (preg_match('#\n \* @final( .*)?\r?\n \*(?: @|/$)#s', $refl->getDocComment())) {
+                self::$final[$name] = true;
+            }
+
+            $parent = get_parent_class($class);
+            if ($parent && isset(self::$final[$parent])) {
+                @trigger_error(sprintf('Extending the class %s in %s is deprecated. This class will be made final in the next major release.', $parent, $name), E_USER_DEPRECATED);
+            }
+
             if (in_array(strtolower($refl->getShortName()), self::$php7Reserved)) {
                 @trigger_error(sprintf('%s uses a reserved class name (%s) that will break on PHP 7 and higher', $name, $refl->getShortName()), E_USER_DEPRECATED);
             } elseif (preg_match('#\n \* @deprecated (.*?)\r?\n \*(?: @|/$)#s', $refl->getDocComment(), $notice)) {
                 self::$deprecated[$name] = preg_replace('#\s*\r?\n \* +#', ' ', $notice[1]);
             } else {
+                // Don't trigger deprecations for classes in the same vendor
```

Comment:
Isn't it? From what I understood, when a class (let's say `App\Controller\AppController`) is deprecated, deprecations won't be deprecated won't be triggered when for example `App\Controller\BlogController` extends it.
Would you prefer:
```php
// Don't trigger deprecations for classes in the same sub namespace
// (e.g. a deprecated class `App\Foo\Deprecated` won't trigger a deprecation
// when extended by `App\Foo\NewClass` but extended by `App\Quz\Bar`, it will).
// For Symfony components/bundles/bridges, this rule applies to the root 
// namespace (e.g. `Symfony`).
```
?

If that locks us I can do it in another PR, it was just the occasion to clarify what this check does ;)